### PR TITLE
add support for including external WMF and EMF image files

### PR DESCRIPTION
**Pull request by warren-bank**
ex:
```text
  {\field {\*\fldinst import image.wmf}}
  {\field {\*\fldinst import image.emf}}
```

notes:
* doesn't directly add support, but does add a hook function and other related plumbing that enables the code that uses the library to perform the async work and data preparation.. then feed it back into the library (via callback)
* the library is able to make proper use of such out-of-band data

comments:
* interesting library
* I spent a little time playing around with it yesterday.. and made these changes
* I'm probably not going to spend any more time on it
* I just wanted to share my thoughts.. and you are, of course, welcome to cherry-pick the ones you like

### CODE REVIEW

**Comment by zoehneto** [comment]
Thank you for your contribution. I just checked the rtf spec (1.9.1) and i can't find a reference to this import feature. Is this supported by other tools or included in an earlier version of the spec?

**Comment by warren-bank** [comment]
That's a very good question; off-hand, I have absolutely no idea..

Anecodally, if I open the [RTF used in sample #4](https://github.com/warren-bank/rtf.js/tree/import-image-support/samples/04_rtf_imports/data) in [Universal Viewer Pro](http://www.uvviewsoft.com/uviewer/), the external images __are__ imported. So, I'd assume that it's a part of the standard..

Googling..
* terms: ["fldinst import"](https://www.google.com/search?q="fldinst%20import")
  * not much
    * links to source code in a handful of projects
* terms: [RTF import external image](https://www.google.com/search?q=rtf+import+external+image)
  * good stuff..
    * it looks like it may be an RTF enhancement added by Microsoft to some of its products
      * MS Word > RTF editor > Insert > "Link Picture"
      * MS Word > RTF editor > Insert > "Link to File"

I'm probably the wrong person to ask :)
It appears that it may possibly be a feature that is common, but not yet standardized?

**Comment by warren-bank** [comment]
[interesting..](https://www.safaribooksonline.com/library/view/rtf-pocket-guide/9781449302047/ch01.html#I_programlisting1_d1e4728)

another alternative and (per the book) "quite nonstandard" way to include an external image:
```text
{\field\fldedit{\*\fldinst { INCLUDEPICTURE  \\d
 "PicturePath"
 \\* MERGEFORMATINET }}{\fldrslt {  }}}
```

**Comment by zoehneto** [comment]
If it is supported by Word I think it qualifies for inclusion. In terms of merging these commits I would prefer it if you could split the changes into the following PRs to improve understanding and review:
- Clean up whitespace (it would probably in this context also make sense to add an [EditorConfig](http://editorconfig.org/)
- apply 0px defaults to width and height
- refactor samples
- add support for including external WMF and EMF image files

Aside from that there are some moved files which were for some reason also changed (jquery.svg.css, simplepict.rtf etc.) and please add some change description to jquery.binarytransport.js so it is obvious that it 's modified.

File: samples/.common/dep/jquery.binarytransport.js [hunk 1]
```diff
@@ -0,0 +1,54 @@
+ /**
+ *
+ * jquery.binarytransport.js
+ *
+ * @description. jQuery ajax transport for making binary data type requests.
```
Comment (zoehneto) [inline samples/.common/dep/jquery.binarytransport.js]: Please add some change description to jquery.binarytransport.js so it is obvious that it 's modified.

File: README.md [hunk 2]
```diff
@@ -15,7 +15,17 @@ The code of this project is licensed under the MIT license.  See the file LICENS
 * Rendering WMF and EMF images is accomplished by using HTML5's `<svg>` feature.
 * Rendering RTF documents often requires rendering embedded WMF and EMF images, however rtf.js can be used without wmf.js and emf.js or with just one of them, if rendering such images is not required.
 
-# Live samples:
-* RTF document rendering: http://tbluemel.github.io/rtf.js/samples/rtf.html
-* WMF image rendering: http://tbluemel.github.io/rtf.js/samples/wmf.html
-* EMF image rendering: http://tbluemel.github.io/rtf.js/samples/emf.html
+# Live samples (original):
```
Comment (zoehneto) [inline README.md]: This should be combined / updated and point to tbluemel/rtf.js before merging.

**Comment by warren-bank** [comment]
* the update to `jquery.binarytransport.js` is unrelated to importing external files.. it's a [one-line change](https://github.com/tbluemel/rtf.js/pull/29/commits/3f92aa5156a209a8da6bba0c909254e4372c1b23) to support running these tests in the browser locally w/ `file://` URIs
  * I added it for my own sanity, because I was testing the samples locally w/o bothering to start a httpd (at localhost)
  * however, for this to work.. also requires starting Chrome w/ a special flag (`--allow-file-access-from-files`) to allow XHR requests to the local file system

* the real meat-and-potatoes of this PR can all be found in [this single commit](https://github.com/tbluemel/rtf.js/pull/29/commits/52741aa3f4924f189cedb43302a3e1c155804624)
  * I'm not sure why some files that were moved are marked as changed..
    * this commit doesn't contain any corrections for whitespace, etc
    * if the files look identical, then they are.. and you can safely ignore them (ie: keep the current ones)

* as for re-organizing the "samples" directory..
  * I just did that for my own sanity
  * if you like the current way that things are organized.. then you can safely ignore the refactoring

File: rtf.js [hunk 3]
```diff
@@ -31,12 +31,18 @@ if (typeof RTFJS === "undefined") {
 			this.message = message;
 			this.stack = (new Error()).stack;
 		},
-		loggingEnabled: true,
```
Comment (zoehneto) [inline rtf.js]: Please leave the default as is, if someone doesn't want logging they can set it to false after creating the RTFJS object.

File: rtf.js [hunk 4]
```diff
@@ -31,12 +31,18 @@ if (typeof RTFJS === "undefined") {
...
+		loggingEnabled: false,
 		log: function(message){
 			if(RTFJS.loggingEnabled) {
 				console.log(message);
 			}
 		},
+		warningEnabled: true,
```
Comment (zoehneto) [inline rtf.js]: This property and the corresponding function are never used.

**Comment by zoehneto** [comment]
I'd like to merge all these changes but I can't merge them as they are right now (to many different unrelated changes in one PR). The commit you linked also requires the sample restructuring to work and a README change (so people can find the example). Aside from that I've also added comments to the rtf.js changes.

**Comment by warren-bank** [comment]
you're right about logging..

I added `RTFJS.warningEnabled` and `RTFJS.warn` to debug a few things.. which I guess were all ultimately deleted prior to commit.  They can both be removed.. and `RTFJS.loggingEnabled` left active.

I apologize.. I suppose that I should've cleaned up the code a bit before sharing.
I'd definitely be willing to do that.. and break up the changes as you suggested.

**Comment by warren-bank** [comment]
..I'll do that within the next day or so

**Comment by warren-bank** [comment]
dumb question..

if the first PR that I do is to clean up white-space, is it right to assume that you prefer tabs? It was kind of a 60/40 mix.. that seemed to favor tabs, but I could do whichever you like best.

**Comment by zoehneto** [comment]
Good question actually :) I think @tbluemel originally formatted with tabs and over time it eroded (probably because my IntelliJ really doesn't like tabs). So I think we should do 4 spaces and also create an [EditorConfig](http://editorconfig.org/) with that setting so we can avoid these issues in the future.

**Comment by warren-bank** [comment]
agreed.. I always do 4 spaces also, but.. not my project :)

**Comment by warren-bank** [comment]
could just borrow the [.editorconfig](https://github.com/facebook/jest/blob/master/.editorconfig) from Jest.. and change it from 2 spaces to 4

**Comment by zoehneto** [comment]
Looks good

**Comment by zoehneto** [comment]
Closing this since it is now handled in PR #32

**Pull request closed without merging**